### PR TITLE
update releases, backup release selection, activate monitoring service

### PR DESCRIPTION
**Pull request by ogmueller**
added tags to tasks to ease filtering

added backup selection of a distribution release, if current release is not available yet. problem here is, that hwraid website doesn't update this information very fast

fixed selection of package. now using hwraid_register_release intead of given hwraid_release, because hwraid_register_release will be available, but hwraid_release might not

changed package installation to present instead of latest, because it should not the task of ansible to update the server, but rather put it into a certain state

added raid monitoring daemon to get started so people get informed, if raid fails

updated available distribution releases for debian and ubuntu as stated on hwraid website

### CODE REVIEW

**Comment by drybjed** [comment]
Thanks! I'll create the test for this role shortly, to make sure it passes on Travis.

I would remove the tags from the tasks. The whole role is tagged in the DebOps playbook as `role::hwraid`, you can use this tag to call the role instead of running separate tasks.

**Comment by ogmueller** [comment]
there is just one problem i had with this. maybe you know a solution to it.
when this role finds a new/better release of hwraid, than the apt-source.d file contains 2 or more versions all of the sudden. like this:

# /etc/apt/sources.list.d/hwraid_le_vert_net_ubuntu.list
deb http://hwraid.le-vert.net/ubuntu trusty main
deb http://hwraid.le-vert.net/ubuntu precise main

**Comment by drybjed** [comment]
Not sure how to solve the above problem yet.

Since you started working on this role, can you move all of the `vars/main.yml` variables to `defaults/main.yml`? That way they will be modifable from the Ansible inventory if needed.

**Comment by drybjed** [comment]
If you want, you could also rename all of the role variables so that they are in their own namespace. Rename `hwraid_` to `hwraid__` (two underscores).

**Comment by ogmueller** [comment]
that i don’t understand. is that neccessery or common practice?

my comment about the apt repository was just, that if you used hwraid already with e.g. precise and you select trusty now, it will create an apt repository file, which has 2 entries in it. i don’t know, if it is possible to prevent easily.

**Comment by ogmueller** [comment]
ok, i moved the vars to defaults

**Comment by drybjed** [comment]
You mean the "namespacing" of the variables? I don't think it's a common practice outside of DebOps. I started doing that so that longer, multiple-word variables are easily readable. Consider this variable name:
```
apt_cacher_ng_ferm_dependent_rules
```
Would you know from the name, that this variable is for the `debops.apt_cacher_ng` role, and defines configuration for `debops.ferm` role? How about this name?
```
apt_cacher_ng__ferm__dependent_rules
```
This should make it more easily recognizable.

Ansible has only one "level' of variables, unfortunately, no real namespaces. You could put all of the role variables in some kind of global dictionary and separate them that way, but to change one thing you would need to copy all of the variables to Ansible inventory. Putting each variable separately makes this easier, but introduces massive amount of variables on that first level, and after a while creating sensible variable names becomes cumbersome. This artifical namespacing with two underscores should help with that problem.

As for the `apt_repository` issue - why did you change the distribution release in the first place? Stick to the one your host has. The problem could be solved if you tracked what distribution release was enabled, disabled the old one and enabled the new one. But, since you should just use the release your OS uses, that becomes pointless.

**Comment by drybjed** [comment]
Changes look good, and I would still remove the tags in the role itself.

**Comment by ypid** [comment]
@drybjed About the name scpacing, note that you setup a wiki which you can refer to for explaining the intention behind this :wink: https://wiki.debops.org/roadmap#project-wide_changes

**Comment by drybjed** [comment]
@ypid Yeah, I know. I guess the main `debops.org` webpage needs to be switched to it at some point. :-)

File: defaults/main.yml [hunk 1]
```diff
@@ -9,3 +9,65 @@ hwraid_release: '{{ ansible_distribution_release }}'
 # Ignore specified kernel modules
 hwraid_blacklist: []
 
+
+
+# HWRaid repository GPG key
+hwraid_repository_key_url: 'http://hwraid.le-vert.net/debian/hwraid.le-vert.net.gpg.key'
```
Comment (ypid) [inline defaults/main.yml]: @ogmueller If you want to, maybe you can change the HTTP URL to the key fingerprint? See: https://github.com/jchaney/owncloud/pull/12 and https://github.com/debops/ansible-owncloud/blob/370d90ebd6b6bc96be72cc0a8bb497f3f3759795/defaults/main.yml#L83-L86. When I downloaded the fingerprint, it was: 0073C11919A641464163F7116005210E23B3D3B4. Maybe that can be crosschecked.

Comment (drybjed) [inline defaults/main.yml]: Definitely a good idea, taking the GPG key from a different source than the APT repository information helps with authentication of the keys. :+1:

**Comment by ypid** [comment]
> I would remove the tags from the tasks. The whole role is tagged in the DebOps playbook as role::hwraid, you can use this tag to call the role instead of running separate tasks.

Tags might still come in handy in tasks, but note that tags are namespaced in DebOps. See [Ansible tags for `debops.bootstrap`](http://docs.debops.org/en/latest/ansible/roles/ansible-bootstrap/docs/getting-started.html#ansible-tags) as an example.

**Comment by drybjed** [comment]
Tags are useful in larger, more complex roles, but with `debops.hwraid` they would only clutter the Ansible output and not be very useful, let me explain why:

First of all, let's see the current output of `debops service/hwraid --list-tasks`:
```
  play #1 (debops_service_hwraid,debops_hwraid): Configure HWRaid support       TAGS: []
    tasks:
      debops.hwraid : Configure HWRaid APT repository key       TAGS: [config, hwraid, package, role::hwraid, system]
      debops.hwraid : Select supported release for current distribution TAGS: [hwraid, install, package, role::hwraid, system]
      debops.hwraid : Select latest release for current distribution if no match found  TAGS: [hwraid, install, package, role::hwraid, system]
      debops.hwraid : Configure HWRaid APT repository   TAGS: [hwraid, install, package, role::hwraid, system]
      debops.hwraid : Get list of active kernel modules TAGS: [hwraid, install, package, role::hwraid, system]
      debops.hwraid : Install packages for recognized RAID devices      TAGS: [hwraid, install, package, role::hwraid, system]
      debops.hwraid : Make sure service starts at boot  TAGS: [hwraid, install, package, role::hwraid, system]
```
As you can see, ale of the tasks have the same tags. So even if you would want to run the tasks tagged with `install`, you would run all of the role tasks. The same with other tags like `package`, `system`, etc. So the tasks in the role become redundant; the whole role in the playbook is already tagged with `role::hwraid`. Or, you can just run the playbook itself directly, which would accomplish the same, since only 1 role is present.

Using different tags in different tasks could make sense, until you realize that you still need to run basically all tasks for the install task to work. So, running only a subset of the role with just 7 tasks will only slow you down since you would need to look up what tags are there to use. Just run the whole role/playbook, it's designed to be idempotent anyway. :-)

**Comment by ypid** [comment]
> Using different tags in different tasks could make sense, until you realize that you still need to run basically all tasks for the install task to work. So, running only a subset of the role with just 7 tasks will only slow you down since you would need to look up what tags are there to use. Just run the whole role/playbook, it's designed to be idempotent anyway. :-)

You are right. I just like to use tags for role development/debuggung as it gets you a few seconds of run time :wink:

**Comment by drybjed** [comment]
@ypid I use temporary `debug` tag for that while developing, but usually only when there's an hard to trace error somewhere later in the role.

**Comment by andyshinn** [comment]
What is the status of this? I'd love to have an Ansible Galaxy version with at least the updated `hwraid_distribution_releases` for Ubuntu 14.04.

**Comment by drybjed** [comment]
I'll merge this, however I'll remove the task tags since they are redundant.

**Comment by ogmueller** [comment]
sorry for the late response. i would actually keep the tags. yes, they are superfluous for you or for people who are using the debops playbooks, but for those guys (like me) who are only using the role only, the tags are not superfluous, but rather essential.

**Comment by drybjed** [comment]
Presumably you cannot use the role by itself, but in an Ansible playbook. In that case all you need to do to achieve the same effect as tagging all of the tasks inside the role with tags, is to write in your playbook:
```
---
- hosts: [ 'debops_service_hwraid' ]
  become: True

  roles:
    - role: debops.hwraid
      tags: ['hwraid', 'install', 'package', 'system' ]
```
In your PR, all of the role tasks were tagged with the same set of 4 tags, so specifying just one of them would trigger all of the role tasks anyway. Tagging the entire role on a playbook level makes more sense in this case.

Ansible tags are only useful when your playbook contains multiple roles with many tasks, and you might want to re-run only a specific set of tasks, for example to change the configuration without all of the other steps. This is commonly done in DebOps roles with tags written in a special format:
```
role::<rolename>:<tagname>
```
This format allows to specify exactly which tasks in which role should be executed. This might be important when a playbook uses multiple roles - using the same tag name, for example `packages` would trigger multiple tasks from different roles.

For example, `debops.ferm` role has tasks tagged with:
```
role::ferm:rules
```
This allows to apply `ferm` rules in any playbook that uses `debops.ferm` role.

However in `debops.hwraid` case, it uses only a few tasks (I want to reduce that number even further soon, by moving some logic to default variables). Almost all of them need to be run each time to make sense, withou `set_fact` tasks other tasks will break. In this case it's hard to create a sensible set of tags for a subset of role tasks, and tagging all tasks of a role should be done on a playbook level to allow for easier modification of the tag list.

**Pull request merged**